### DISCUSSION TITLE
Generating different APIs from the same schema

### DISCUSSION THREAD

**Discussion by eli5zen** [supabase/supabase#7542] [Feature Requests]
Hey :)

I have a use case that may be relevant for more people (maybe a similar form of the problem) so after talking to the support team, they advised me to open up a discussion here :)

## TL;DR
Generate multiple APIs on top of the same schema (and DB), each API exposing parts of the schema.

## Detailed
We develop 2 applications. One of them is B2C, which will be open-sourced & free to use. The other is B2B, closed & has costs for usage. The apps share & interact on some of the data, so to enforce consistent and accurate relationships on both sides, we wish to use a single PostgreSQL DB. Can supabase help with generating 2 sets of APIs from the same DB (so private schema is not exposed in the open-sourced API)?

See below a diagram for clarification:

<img width="387" alt="Screen Shot 2022-07-03 at 22 10 16" src="https://user-images.githubusercontent.com/10067209/177053992-76fae95f-60e3-43b7-97dd-8af2b854066d.png">

Thanks!

**Comment by silentworks** [comment]
Supabase provides you with an API based on your database schema, but the access control on these would be totally up to you the developer. You can use row level security (RLS) to restrict permissions to any specific table based on a query or custom claims. You can also create your own API SDK using any framework/language you wish and just use Supabase as a Postgres database.

These are the two approaches I could think of:

1. Use RLS to restrict access to the tables in the private schema based on user credentials or a custom claim. This will allow you to use the supabase-js library to build out both parts of the system.
2. Create a custom API SDK with your language/framework of choice, enable RLS on all tables in the private schema with no policy attached to them (by default no one can access them with the supabase-js library if RLS is on and no policy is set). This will allow you to use the supabase-js library for the open-source part of the system while using your custom API SDK for the private part.

**Comment by burggraf** [comment]
Here's one way to do custom claims (for item 1) if you need it: [supabase-community/supabase-custom-claims](https://github.com/supabase-community/supabase-custom-claims)

**Comment by steve-chavez** [comment]
> Generate multiple APIs on top of the same schema (and DB), each API exposing parts of the schema.

You could also consider splitting your schema into two, you can move a table to another schema with `ALTER TABLE existing SET SCHEMA b2c` .

Another choice that wouldn't touch your existing tables. Create two schemas and make views of your existing tables on those.

```sql
create schema b2b;
create schema b2c;

create or replace view b2b.existing as
select * from private.existing;

create or replace view b2c.another as
select * from private.another;
```

Then grant usage on those schemas to different/isolated roles.(`GRANT USAGE ON SCHEMA b2c TO b2c_role; GRANT USAGE ON SCHEMA SCHEMA b2b TO b2b_role;`).
